Source/WebCore/dom/InternalObserverToArray.cpp

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2025 Marais Rossouw <me@marais.co>. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "InternalObserverToArray.h"
+
+#include "InternalObserver.h"
+#include "JSDOMPromiseDeferred.h"
+#include "JSValueInWrappedObject.h"
+#include "Observable.h"
+#include "ScriptExecutionContext.h"
+#include "SubscribeOptions.h"
+#include <wtf/Vector.h>
+
+namespace WebCore {
+
+class InternalObserverToArray final : public InternalObserver {
+public:
+    static Ref<InternalObserverToArray> create(ScriptExecutionContext& context, Ref<DeferredPromise>&& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverToArray(context, WTFMove(promise)));
+        internalObserver->suspendIfNeeded();
+        return internalObserver;
+    }
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        auto* globalObject = protectedScriptExecutionContext()->globalObject();
+        ASSERT(globalObject);
+
+        Ref vm = globalObject->vm();
+
+        m_list.append({ vm, value });
+    }
+
+    void error(JSC::JSValue value) final
+    {
+        m_promise->reject<IDLAny>(value);
+    }
+
+    void complete() final
+    {
+        InternalObserver::complete();
+
+        m_promise->resolve<IDLSequence<IDLAny>>(m_list);
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor&) const final
+    {
+    }
+
+    InternalObserverToArray(ScriptExecutionContext& context, Ref<DeferredPromise>&& promise)
+        : InternalObserver(context)
+        , m_promise(WTFMove(promise))
+    {
+    }
+
+    Vector<JSC::Strong<JSC::Unknown>> m_list;
```

Comment:
Let me try to be even more explicit than @cdumez. I think this may be incorrect and will create a permanent leak if there are reference cycles. We should create a test that proves we can collect even if you try to make a reference cycle that intentionally covers the most difficult edge case; it's possible we can't use `Strong` and pass such a test.

Reply:
Great call out!! 🤩 Ill have to figure out where and how to write that test. Hypothetically (maybe even more) what would an alternate approach be should the test prove this leaks?

Reply:
An alternate approach needs to put the items where they will be marked during garbage collection as a side effect of marking the `InternalObserverToArray` object. I don’t know offhand all the details of how that would be done are; it can be tricky to get right because of the garbage collection happening on another thread.

Reply:
One thing that might work would be using a temporary JavaScript array object to hold the list, and then have a reference to that JavaScript array be stored in a `JSValueInWrappedObject`.